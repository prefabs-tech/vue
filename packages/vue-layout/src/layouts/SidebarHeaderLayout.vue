<template>
  <div
    :aria-expanded="appHeader?.expanded"
    :class="['layout sidebar-header-layout', { collapsible: collapsible }]"
  >
    <slot name="header">
      <AppHeader
        ref="appHeader"
        :no-locale-switcher="noLocaleSwitcher"
        :no-toggle="!collapsible"
        no-main-menu
      >
        <template v-if="userMenuLocation === 'header'" #userMenu>
          <slot name="userMenu"></slot>
        </template>
        <template #locales>
          <slot name="locales" />
        </template>
      </AppHeader>
    </slot>
    <Sidebar
      ref="sidebar"
      :collapsible="collapsible"
      :menu="menu"
      :no-header="noSidebarHeader"
      class="layout-sidebar"
    >
      <template #afterNavLinks>
        <slot name="afterNavLinks"></slot>
      </template>

      <template #afterSidebarMenu>
        <slot name="afterSidebarMenu"></slot>
        <slot v-if="userMenuLocation === 'sidebar'" name="userMenu"></slot>
        <template v-if="sidebarLocaleSwitcher">
          <slot name="locales">
            <LocaleSwitcher v-if="!noLocaleSwitcher" class="locales" />
          </slot>
        </template>
      </template>

      <template #footer>
        <slot name="footer">
          <AppFooter />
        </slot>
      </template>
    </Sidebar>
    <main>
      <slot></slot>
    </main>
  </div>
</template>

<script setup lang="ts">
import { LocaleSwitcher } from "@prefabs.tech/vue3-i18n";
import { ref, watch, onUnmounted } from "vue";

import AppFooter from "../components/AppFooter.vue";
import AppHeader from "../components/AppHeader.vue";
import Sidebar from "../components/Sidebar.vue";

import type { SidebarMenu } from "../types";
import type { PropType } from "vue";

const appHeader = ref();
const sidebar = ref();
const sidebarLocaleSwitcher = ref<boolean>(false);

defineProps({
  collapsible: {
    default: true,
    type: Boolean,
  },
  menu: {
    required: true,
    type: Array as PropType<SidebarMenu[]>,
  },
  noLocaleSwitcher: Boolean,
  noSidebarHeader: {
    default: false,
    type: Boolean,
  },
  userMenuLocation: {
    default: "sidebar",
    type: String,
    validator: (value: string) => ["header", "sidebar"].includes(value),
  },
});

watch(
  () => appHeader.value?.expanded,
  (newValue) => {
    if (newValue) {
      sidebar.value.sidebarActive = true;
    }
  },
);

watch(
  () => sidebar.value?.sidebarActive,
  (newValue) => {
    if (!newValue) {
      appHeader.value.expanded = false;
    }
  },
);

const handleResize = () => {
  if (window.innerWidth >= 576) {
    sidebarLocaleSwitcher.value = false;
  } else {
    sidebarLocaleSwitcher.value = true;
  }
};

onUnmounted(() => {
  window.removeEventListener("resize", handleResize);
});

window.addEventListener("resize", handleResize);

handleResize();
</script>

<style lang="css">
@import "../assets/css/layouts/sidebarHeader.css";
</style>
