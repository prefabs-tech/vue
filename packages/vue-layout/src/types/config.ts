import type { MenuItem } from "./menu";
import type { AppConfig } from "@prefabs.tech/vue3-config";
import type { LocaleMessages, VueMessageType } from "@prefabs.tech/vue3-i18n";

interface PrefabsTechVue3LayoutConfig {
  homeRoute?: "home" | string;
  localeSwitcher: {
    showBadges: true,
  },
  logo?: string;
  logoAlt?: string;
  logoRoute?: string;
  mainMenu?: MenuItem[];
}

interface PrefabsTechVue3LayoutPluginOptions {
  config: AppConfig;
  translations?: LocaleMessages<VueMessageType>;
}

export type { PrefabsTechVue3LayoutConfig, PrefabsTechVue3LayoutPluginOptions };
