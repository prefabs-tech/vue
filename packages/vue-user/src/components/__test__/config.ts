import type { AppConfig } from "@prefabs.tech/vue3-config";
import type { LocaleMessages, VueMessageType } from "@prefabs.tech/vue3-i18n";

const config: AppConfig = {
  apiBaseUrl: "http://localhost",
  appTitle: "Test app",
  appVersion: "0.1.0",
  i18n: {
    fallbackLocale: "en",
    messages: {} as LocaleMessages<VueMessageType>,
    supportedLocales: ["en", "fr"],
  },
  features: {
    showVersion: true,
  },
  sentry: {
    dsn: "tbd",
  },
  slug: "app",
  user: {
    features: {
      confirmPassword: true,
    }
  },
  websiteDomain: "test",
};

export default config;
