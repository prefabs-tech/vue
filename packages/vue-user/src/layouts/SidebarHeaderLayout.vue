<template>
  <Layout
    :collapsible="collapsible"
    :menu="menu"
    :no-locale-switcher="noLocaleSwitcher"
    :user-menu-location="userMenuLocation"
    no-sidebar-header
  >
    <template #header>
      <slot name="header"></slot>
    </template>

    <template #locales>
      <slot name="locales"></slot>
    </template>

    <template #userMenu>
      <UserMenu v-if="showUserMenu">
        <template v-if="$slots.userMenuTrigger" #userMenuTrigger>
          <slot name="userMenuTrigger"></slot>
        </template>
      </UserMenu>
    </template>

    <template #afterNavLinks>
      <slot name="afterNavLinks"></slot>
    </template>

    <template #afterSidebarMenu>
      <slot name="afterSidebarMenu"></slot>
    </template>

    <template #footer>
      <slot name="footer"></slot>
    </template>

    <slot></slot>
  </Layout>
</template>

<script lang="ts">
export default {
  name: "SidebarHeaderLayout",
};
</script>

<script setup lang="ts">
import { useConfig } from "@prefabs.tech/vue3-config";
import { SidebarHeaderLayout as Layout } from "@prefabs.tech/vue3-layout";
import { storeToRefs } from "pinia";
import { computed } from "vue";
import { useRouter } from "vue-router";

import UserMenu from "../components/UserMenu.vue";
import useUserStore from "../store";

import type { MenuItem, SidebarMenu } from "@prefabs.tech/vue3-layout";

defineProps({
  collapsible: {
    default: true,
    type: Boolean,
  },
  noLocaleSwitcher: Boolean,
  showUserMenu: {
    default: true,
    required: false,
    type: Boolean,
  },
  userMenuLocation: {
    default: "sidebar",
    type: String,
    validator: (value: string) => ["header", "sidebar"].includes(value),
  },
});

const userStore = useUserStore();

const { user } = storeToRefs(userStore);

const { layout: layoutConfig } = useConfig();

const router = useRouter();

const allRoutes = router.getRoutes();

const menu = computed(() => {
  let menuItems = layoutConfig?.mainMenu || [];

  if (!user.value) {
    menuItems = menuItems?.filter((item: MenuItem) => {
      const route = allRoutes.find((r) => {
        if (r.name === item.route && !r.meta?.authenticated) {
          return true;
        }

        if (item.children?.length) {
          return (item.children as MenuItem[]).some((child: MenuItem) => {
            return allRoutes.some(
              (cr) => cr.name === child.route && !cr.meta?.authenticated,
            );
          });
        }

        return false;
      });

      return !!route;
    }) as MenuItem[];
  }

  return menuItems.map((item: MenuItem) => {
    return {
      hide: item?.hide,
      icon: item?.icon,
      name: item.name,
      routeName: item.route,
      shortName: item?.shortName,
      children: item?.children?.map((childItem: MenuItem) => {
        return {
          hide: childItem?.hide,
          icon: childItem?.icon,
          name: childItem.name,
          routeName: childItem.route,
          shortName: childItem?.shortName,
        };
      }),
    };
  }) as SidebarMenu[];
});
</script>

<style lang="css">
.sidebar-header-layout > .sidebar .user-menu > ul {
  flex-direction: column;
}

.sidebar-header-layout > .sidebar .user-menu > ul > li,
.sidebar-header-layout > .sidebar .user-menu-dropdown {
  --_font-size: var(--font-size-min, 0.8rem);
  --_font-weight: var(--font-weight, 450);
  --_layout-sidebar-separator-color: var(
    --layout-sidebar-separator-color,
    #dbdbdb
  );

  font-size: var(--_font-size);
  font-weight: var(--_font-weight);
  width: 100%;
}

.sidebar-header-layout
  > .sidebar
  .user-menu
  > ul
  > li:has(.router-link-exact-active) {
  background-color: var(--nav-item-active-bg, #eff6ff);
  color: var(--nav-item-active-color, #2c2c2c);
}

.sidebar-header- > .sidebar .user-menu:not(.user-menu-dropdown) > ul > li > a,
.sidebar-header-layout > .sidebar .user-menu-dropdown,
.sidebar-header-layout > .sidebar .user-menu-dropdown > ul.dropdown > li {
  align-items: center;
  display: flex;
  width: 100%;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown {
  border-top: 1px solid var(--_layout-sidebar-separator-color);
  padding-block: 0.5rem;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown,
.sidebar-header-layout > .sidebar .user-menu-dropdown:hover {
  background-color: #fff;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown > .trigger {
  background-color: var(--sidebar-bg-color, #fff);
  width: 100%;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown > .trigger:hover {
  background-color: #f0f2f7ad;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded > ul > li:hover {
  --dropdown-bg-color-hover: #f0f2f7ad;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded .toggle > svg {
  transform: rotate(-180deg);
}

.sidebar-header-layout > .sidebar .user-menu-dropdown .toggle > svg {
  transform: rotate(0);
}

.sidebar-header-layout > .sidebar .user-menu-dropdown > .dropdown {
  transform: translate3d(0, -80px, 0);
  transition:
    transform 0.2s ease,
    visibility 0.1s ease;
  visibility: hidden;
  width: 100%;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded > .dropdown {
  --dropdown-bg-color: #fff;

  box-shadow: 0 -2px 12px #0000001a;
  transform: translate3d(0, -161.5px, 0);
  visibility: visible;
  width: 100%;
}
</style>
