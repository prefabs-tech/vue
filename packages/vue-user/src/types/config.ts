import { DefineComponent } from "vue";

import type { RouteOverrides } from "./router";
import type {
  IsEmailOptions,
  StrongPasswordOptions,
} from "@prefabs.tech/vue3-form";

interface DzangolabVueUserConfig {
  apiRoutes?: {
    changeEmail?: string;
    changePassword?: string;
    getVerificationStatus?: string;
    login?: string;
    logout?: string;
    passwordReset?: string;
    passwordResetRequest?: string,
    refresh?: string;
    sendVerificationEmail?: string;
    signup?: string;
    verifyEmail?: string;
  };
  features?: {
    authProvider?: string;
    forgotPasswordResendTimeInSeconds?: number;
    loginType?: "email" | "username";
    signUp?: {
      emailVerification?: boolean;
      termsAndConditions?: {
        display?: boolean;
        route?: string;
        showCheckbox?: boolean;
      };
    };
    updateEmail?: boolean;
  };
  password?: {
    minLength: number;
  };
  options?: {
    email?: IsEmailOptions;
    password?: StrongPasswordOptions;
  };
  routes?: RouteOverrides;
  socialLogins?: string[];
  supportedRoles?: string[];
}

declare module "@prefabs.tech/vue3-config" {
  export interface AppConfig {
    user?: DzangolabVueUserConfig;
  }
}

export type { DzangolabVueUserConfig };
