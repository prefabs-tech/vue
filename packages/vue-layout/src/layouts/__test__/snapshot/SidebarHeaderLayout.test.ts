import configPlugin from "@prefabs.tech/vue3-config";
import { shallowMount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import config from "../../../components/__test__/config";
import SidebarHeaderLayout from "../../SidebarHeaderLayout.vue";

import type { VueWrapper } from "@vue/test-utils";

describe("SidebarHeaderLayout", () => {
  it("matches snapshot", () => {
    const wrapper: VueWrapper = shallowMount(SidebarHeaderLayout, {
      global: {
        plugins: [
          [
            configPlugin,
            {
              config,
            }
          ],
        ],
      },
      props: {
        menu: [
          {
            name: "Menu1",
            routeName: "menu-1",
            shortName: "M1",
          },
          {
            name: "Menu2",
            routeName: "menu-2",
            shortName: "M2",
          },
        ],
      },
      slots: {
        default: "<div class='default'></div>",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
