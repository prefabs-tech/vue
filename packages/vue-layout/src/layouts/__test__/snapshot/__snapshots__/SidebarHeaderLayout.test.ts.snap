// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarHeaderLayout > matches snapshot 1`] = `
<div
  class="layout sidebar-header-layout collapsible"
>
  
  <app-header-stub
    nolocaleswitcher="false"
    nologo="false"
    nomainmenu="true"
    notoggle="false"
  />
  
  <sidebar-stub
    class="layout-sidebar"
    collapsible="true"
    menu="[object Object],[object Object]"
    noheader="false"
  />
  <main>
    
    <div
      class="default"
    />
    
  </main>
</div>
`;
